server/src/main/java/timely/sample/iterators/AggregationIterator.java

```diff
@@ -0,0 +1,167 @@
+package timely.sample.iterators;
+
+import org.apache.accumulo.core.client.IteratorSetting;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iterators.WrappingIterator;
+import org.apache.commons.lang3.StringUtils;
+import timely.model.Tag;
+import timely.sample.Aggregation;
+import timely.sample.Aggregator;
+import timely.sample.Downsample;
+import timely.sample.Sample;
+
+import java.io.*;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This iterator will aggregate across series as output from the
+ * DownsampleIterator. The same set of aggregator functions are supported as in
+ * the DownsampleIterator.
+ */
+public class AggregationIterator extends WrappingIterator {
+
+    // To keep the output of the aggregation iterator consistent with that of
+    // the downsample iterator, we
+    // supply a static set of tags to be returned with the Aggregation (subclass
+    // of Downsample)
+    private static final String TAGS = "aggregation.tags";
+    private static final String AGGCLASS = "aggregation.aggclass";
+
+    private Aggregation aggregation;
+    private Set<Tag> tags;
+    private Key last;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void init(SortedKeyValueIterator<Key, Value> source, Map<String, String> options, IteratorEnvironment env)
+            throws IOException {
+        super.init(source, options, env);
+        String aggClassname = options.get(AGGCLASS);
+        Class<? extends Aggregator> aggClass;
+        try {
+            aggClass = (Class<? extends Aggregator>) this.getClass().getClassLoader().loadClass(aggClassname);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+        tags = deserializeTags(options.get(TAGS));
+        try {
+            aggregation = new Aggregation(aggClass.newInstance());
+        } catch (InstantiationException | IllegalAccessException e) {
+            throw new RuntimeException("Error creating aggregator class: " + aggClass, e);
+        }
+    }
+
+    @Override
+    public boolean hasTop() {
+        // for each set of downsamples (one per subquery)
+        while (super.hasTop()) {
+            last = super.getTopKey();
+
+            // decode the set of samples (or series)
+            Map<Set<Tag>, Downsample> samples = null;
+            try {
```

Comment:
What I am missing is the possible performance increases gained by avoiding multiple try-catch blocks.  Perhaps minimal but measurable in any case.